core/modules/widgets/range.js

```diff
@@ -111,15 +206,17 @@ Selectively refreshes the widget if needed. Returns true if the widget or any of
 */
 RangeWidget.prototype.refresh = function(changedTiddlers) {
 	var changedAttributes = this.computeAttributes();
-	if(changedAttributes.tiddler || changedAttributes.field || changedAttributes.index || changedAttributes['min'] || changedAttributes['max'] || changedAttributes['increment'] || changedAttributes["default"] || changedAttributes["class"] || changedAttributes.disabled) {
+	if(	changedAttributes.tiddler || changedAttributes.field || changedAttributes.index || changedAttributes['min'] || changedAttributes['max'] ||
```

Comment:
I don't mind the condition being split to multiple lines, but we wouldn't usually do it by adding whitespace between the opening parenthesis and the condition.

Reply:
Using tab characters for alignment (the "if(" line is really "if(→" where → represents a tab character) is a bad idea, because it ensures that the code will only be aligned if someone else uses tab settings that are the same size as yours, thereby removing the one big advantage of tabs for indentation (that people can set their own tab settings). Here you have the right idea by putting a tab after the "if(", so that people with either 4-space or 8-space tab settings will see the code correctly aligned. But if someone has their tab size set to 2 spaces, then this is no longer aligned.

If you are using tabs for indentation, the only correct approach is "tabs for indentation, spaces for alignment". The second line of the `if(` condition should be indented the same level as the `if(` (here that's two tab characters for two indentation levels), with three spaces after the two tab characters to line up the `changedAttributes` text. I.e., the second line should look like `→→···changedAttributes['increment']`, where → represents a tab character and · represents a space. That is the only correct way to do alignment when you have used tabs for indentation, because it ensures that the aligned lines will stay aligned in everyone's editor no matter what tab settings they use.

And the big _problem_ with "tabs for indentation, spaces for alignment" is that most editors are not smart about it, and will mess it up. Which means that if you've aligned your code, you then have to pay careful attention to your whitespace every time you edit that line. Which is a lot of fiddly work that I, for one, don't care for, and that's why my personal preference is to use space characters, not tabs, for indentation.

There are really only four options:
1. Use tab characters naively, for both indentation and alignment, and end up with code that looks different when different editor settings are applied.
1. Use tab characters intelligently, using tabs for indentation and spaces for alignment, and end up paying far more attention to whitespace than it's worth.
1. Use tab characters for indentation and forget trying to align subsequent lines. The rule for function parameters, second lines of "if()" conditions, and so on, is "just indent one more level with one extra tab character". This will still look reasonably good on everyone's editor, and everyone can use their favorite indentation level by tweaking their editor's tab settings.
1. Use space characters for indentation. This allows you to align function parameters, "if()" conditions, and so on, but other coders won't get to use their favorite indentation level and must conform to the indentation level the project has chosen.

My personal preference is for 4. 1 is just bad, 2 causes you to pay way more attention to whitespace than it's worth, and 3 sacrifices alignment, which is a useful tool in many situations. So I'd rather choose option 4 and sacrifice the ability to use my preferred indentation, instead using whatever the project has settled on.

Point is, it's up to you whether you choose 2 or 3 here (4 is not an option since the TiddlyWiki project uses tabs for indentation), but please don't choose option 1. To choose option 2, write your continuation lines as `→→···condition`. To choose option 3, write your continuation lines as `→→→condition`.

Reply:
> 4. Use space characters for indentation. This allows you to align function parameters, "if()" conditions, and so on, but other coders won't get to use their favorite indentation level and must conform to the indentation level the project has chosen.

TW doesn't compress or minify the source code. That makes it easy for every user to start to read the code and start to debug it.

That means the empty.html would contain 4 times more indentation spaces as used with tabs. This would considerably increase the basic file size. So atm TW uses tabs for a reason.

Reply:
Indeed, TW is one of the few projects where the fact that tabs take less storage than spaces is actually relevant. Most of the time compression makes the space savings of tab characters irrelevant, but in TW it actually matters. I wasn't actually arguing that TW should change (I did mention that option 4 is not an option in this case), I was just trying to list all available options for dealing with indentation & alignment. I agree that using tabs for TW makes sense since it isn't minified (though it probably is compressed, given that most servers default to gzipping HTML for delivery).

Though it *is* actually a little hard for every user to read the code, since when you download TW in HTML form via the big "Download Empty" button, the HTML form has had to store the code in a way that looks like "{\n\t\t\t\tresults.push(title);\n\t\t\t}" and so on. Granted, anyone who knows enough about Javascript to want to modify that code also probably knows how to do a search-and-replace to turn "\n" into a real newline and "\t" into a real tab. But there is a small barrier to entry; it's not quite the ideal situation that you described. :-)